### PR TITLE
Stabilize const versions of ptr::slice_from_raw_parts and slice::from_raw_parts.

### LINKED ISSUE #67456: Tracking issue for const slice::from_raw_parts_mut (const_slice_from_raw_parts_mut)
Make `slice::from_raw_parts` and `slice::from_raw_parts_mut` a const fn available under a feature flag.
This would require a change in the `ptr` module as well, as slice module just forwards to it.

`slice::from_raw_parts[mut]` is used in alot of places (e.g `slice::from_ref[mut]`, which would get one step closer into constification if `slice::from_raw_parts[mut]` is a const fn.

Here is a little playground to show it's possible:
https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=dd5c506a3e082c619f557d972e9956ff

In order to get this working, the following functions and functionalities need to be constified:
 - [x] `ptr::slice_from_raw_parts` (https://github.com/rust-lang/rust/pull/67462)
 
---

Implementation PR for `ptr::slice_from_raw_parts`: https://github.com/rust-lang/rust/pull/67462

----

Partially stabilized in https://github.com/rust-lang/rust/pull/97522; remaining unstable functions:

```rust
// core::ptr
pub const fn slice_from_raw_parts_mut<T>(data: *mut T, len: usize) -> *mut [T];

// core::slice
pub const unsafe fn from_raw_parts_mut<'a, T>(data: *mut T, len: usize) -> &'a mut [T];

// core::ptr::NonNull
pub const fn slice_from_raw_parts(data: NonNull<T>, len: usize) -> Self
```

### PR DESCRIPTION
**Pull request by tobz**
This doesn't close the tracking issue (#67456) as there's still the mutable variants to consider, but the immutable variants should seemingly be fine to stabilize.

Apologies if this stabilization PR is lacking something: it wasn't immediately clear if there was a different process for partial stabilizations.

### CODE REVIEW

File: library/core/src/slice/raw.rs [hunk 1]
```diff
@@ -133,8 +133,7 @@ pub const unsafe fn from_raw_parts_mut<'a, T>(data: *mut T, len: usize) -> &'a m
 
 // In debug builds checks that `data` pointer is aligned and non-null and that slice with given `len` would cover less than half the address space
 #[cfg(debug_assertions)]
-#[unstable(feature = "const_slice_from_raw_parts", issue = "67456")]
-#[rustc_const_unstable(feature = "const_slice_from_raw_parts", issue = "67456")]
+#[rustc_const_stable(since = "1.61.0")]
```
Comment (tobz) [inline library/core/src/slice/raw.rs:136]: Admittedly, this is a change I'm not 100% sure on: its callers are stable in a non-const context, and yet it was marked `unstable`.  I have no real clue what `unstable` even implies for a non-exported function, so this was mostly a no-thinking-involved mechanical change.

**Comment by tobz** [comment]
Looking at the CI failure, it looks like this actually unwinds a bit to depending on `ptr::from_raw_parts` being const-stable, which makes sense on a surface level but it also seems to power non-const-fns, despite also being marked `unstable`?

It's not clear to me if it can also be made const stable safely? This is probably one of those "wait for help" scenarios, at this point. :)

**Review by dtolnay — COMMENTED** [review]
Anything used in the implementation of a public stable const fn needs to be marked rustc_const_stable.

```console
error: `ptr::metadata::from_raw_parts` is not yet stable as a const fn
    |
260 |     from_raw_parts(data.cast(), len)
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

https://github.com/rust-lang/rust/blob/f5975e56fef4a861e8e06d3bc0f33e2281637917/library/core/src/ptr/metadata.rs#L108-L110

**Comment by tobz** [comment]
> Anything used in the implementation of a public stable const fn needs to be marked rustc_const_stable.

That's reasonable.  Is that something we/I can do in this case?  Admittedly, it's not clear to me.

**Review by dtolnay — CHANGES_REQUESTED** [review]
Yes, you will need to add rustc_const_stable attributes for all the private functions getting called by these public functions. Whether it's something we can do depends on what functions end up needing to get rustc_const_stable added. That would be reviewed after the changes are made.

**Comment by JohnCSimon** [comment]
Ping from triage:
@tobz - can you please address the conflicts and build failures?

FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

**Comment by tobz** [comment]
I'm going to close this PR for now.

Working through adjusting the const stability for the transitive functions landed me at having to do so for the `ptr_metadata` feature.. which is obviously not ready to be const stable yet, and so spending any more time -- either me making the PR pass or someone reviewing it only to also coming to the same conclusion -- wouldn't make much sense.

**Comment by xfix** [comment]
Is there a problem with marking [`std::ptr::from_raw_parts`](https://doc.rust-lang.org/stable/std/ptr/fn.from_raw_parts.html) as const stable? Note that marking a function as const stable doesn't mean having to stabilize it, see https://rustc-dev-guide.rust-lang.org/stability.html#rustc_const_stable.

> This attribute can make sense even on an `unstable` function, if that function is called from another `rustc_const_stable` function.

**Pull request closed without merging**